File: caro_client_export/full_clean_export/backend/engine/generation/generator.py
"""
Main option generator — sweeps the design space, builds options,
validates, scores, deduplicates, ranks, and exports.
"""

import math
import json
from typing import Any, Dict, List, Optional

from .forms import generate_form
from .rooms import rooms_per_floor, build_floor_programme, YT_TYPES, PAD_TYPES
from backend.engine.validation.validator import validate
from backend.engine.validation.rules import get_rules
from backend.engine.scoring.scorer import score_option
from backend.engine.scoring.cost import estimate_cost

# Brand constants
GROUND_H = 4.5
FLOOR_H = 3.2

_PLANNING = get_rules().get("planning", {})

# Design space — expanded for real site (4,213 m² buildable, 97×55m envelope)
DESIGN_SPACE = {
    "forms":          ["BAR", "BAR_NS", "L", "U", "C"],
    "floor_areas":    [600, 700, 770, 850, 950, 1100, 1300, 1500],  # up to max coverage
    "wing_widths":    [13.6, 14.0, 16.1, 18.0],
    "storeys":        [4, 5, 6, 7, 8],
    "yt_rooms":       [80, 90, 100, 110, 120, 140],
    "pad_units":      [20, 25, 30, 35, 40, 50],
    "corridors":      ["double", "single"],
    "outdoor_opts":   ["WEST", "ROOFTOP", "BOTH"],
    "outdoor_depths": [8, 10, 12, 15],
    "foh_pcts":       [0.55, 0.60, 0.65],
}


def build_option(params: Dict[str, Any], option_id: Optional[str] = None) -> Optional[Dict[str, Any]]:
    """Build a single fully-calculated option from a parameter set."""
    form_type = params["form"]
    fa = params.get("target_floor_area_m2", params.get("floor_area", 770))
    ww = params["wing_width_m"]
    storeys = params["storeys"]
    ct = params["corridor_type"]
    yt_target = params["yt_rooms"]
    pad_target = params["pad_units"]
    od_pos = params.get("outdoor_pos", "WEST")
    od_depth = params.get("outdoor_depth_m", 12)

    # Form geometry
    form = generate_form(form_type, fa, ww)
    if form is None:
        return None

    # Rooms per floor
    yt_rpf = rooms_per_floor(form, ct, YT_TYPES)
    pad_rpf = rooms_per_floor(form, ct, PAD_TYPES)

    # Floor counts
    yt_floors = min(math.ceil(yt_target / yt_rpf), storeys - 1) if yt_target > 0 else 0
    pad_floors = min(math.ceil(pad_target / pad_rpf), storeys - 1 - yt_floors) if pad_target > 0 else 0

    actual_yt = min(yt_target, yt_floors * yt_rpf)
    actual_pad = min(pad_target, pad_floors * pad_rpf)
    total_keys = actual_yt + actual_pad
    if total_keys < 50:
        return None

    # Build floors
    build_params = {**params, "yt_rooms": actual_yt, "pad_units": actual_pad,
                    "length_m": form["bounding_L"], "width_m": form["bounding_W"]}
    floors = build_floor_programme(build_params, yt_rpf, pad_rpf)

    actual_floor_count = len(floors)
    fp = form["footprint_m2"]
    gia = fp * actual_floor_count
    height = GROUND_H + (actual_floor_count - 1) * FLOOR_H

    # NIA
    nia = 0
    for fl in floors:
        for r in fl.get("rooms", []):
            rt = YT_TYPES.get(r["type"]) or PAD_TYPES.get(r["type"])
            if rt:
                nia += rt["nia"] * r["count"]

    # Outdoor
    od_ground = od_depth * form["west_facade_m"] if od_pos in ("WEST", "BOTH") else 0
    od_roof = fp * 0.5 if od_pos in ("ROOFTOP", "BOTH") else 0
    od_court = form["courtyard_m2"]
    od_total = od_ground + od_roof + od_court

    # Facade area
    facade_area = form["total_facade_m"] * height

    # Accessible count
    acc = sum(
        r["count"] for fl in floors for r in fl.get("rooms", [])
        if r["type"] in ("Accessible", "AccStudio")
    )

    metrics = {
        "form": form_type,
        "total_keys": total_keys, "yt_rooms": actual_yt, "pad_units": actual_pad,
        "storeys": actual_floor_count,
        "footprint_m2": round(fp), "gia_m2": round(gia),
        "gia_per_key": round(gia / total_keys, 1),
        "nia_m2": round(nia), "nia_gia_pct": round(nia / max(1, gia) * 100, 1),
        "building_height_m": round(height, 1),
        "coverage_pct": round(fp / _PLANNING.get("site_area_m2", 4213) * 100, 1),
        "far": round(gia / _PLANNING.get("site_area_m2", 4213), 2),
        "west_facade_m": form["west_facade_m"],
        "total_facade_m": form["total_facade_m"],
        "total_facade_m2": round(facade_area),
        "outdoor_ground_m2": round(od_ground),
        "outdoor_roof_m2": round(od_roof),
        "outdoor_courtyard_m2": round(od_court),
        "outdoor_total_m2": round(od_total),
        "accessible_count": acc,
        "corridor_type": ct, "outdoor_pos": od_pos,
        "wing_width_m": ww,
        "yt_per_floor": yt_rpf, "pad_per_floor": pad_rpf,
        "bounding_L": form["bounding_L"], "bounding_W": form["bounding_W"],
        "wings": form["wings"],
    }

    # Cost
    cost = estimate_cost(metrics)
    metrics.update(cost)
    metrics["cost_per_key_usd"] = cost["per_key_usd"]

    # Assemble option
    opt = {"params": params, "metrics": metrics, "floors": floors, "form_data": form}

    # Validate
    is_valid, violations, warnings = validate(opt)
    opt["is_valid"] = is_valid
    opt["violations"] = violations
    opt["warnings"] = warnings

    # Score
    score, breakdown = score_option(opt)
    opt["id"] = option_id or f"OPT-{abs(hash(json.dumps(params, sort_keys=True, default=str))) % 1000000:06d}"
    opt["score"] = score
    opt["score_breakdown"] = breakdown
    opt["rank"] = 0

    return opt


def generate_all(max_options: int = 30, include_invalid: bool = False) -> List[Dict[str, Any]]:
    """Sweep the design space and generate all valid unique options."""
    configs = []

    # Sweep 1: Form × floor area (main axis)
    for f in DESIGN_SPACE["forms"]:
        for fa in DESIGN_SPACE["floor_areas"]:
            for ww in DESIGN_SPACE["wing_widths"]:
                configs.append({
                    "form": f, "target_floor_area_m2": fa, "wing_width_m": ww,
                    "storeys": 6, "corridor_type": "double", "yt_rooms": 100,
                    "pad_units": 30, "outdoor_pos": "WEST", "outdoor_depth_m": 12,
                })

    # Sweep 2: Programme variations
    for f in ["BAR", "BAR_NS", "L"]:
        for yt in DESIGN_SPACE["yt_rooms"]:
            for pad in DESIGN_SPACE["pad_units"]:
                configs.append({
                    "form": f, "target_floor_area_m2": 770, "wing_width_m": 14.0,
                    "storeys": 6, "corridor_type": "double", "yt_rooms": yt,
                    "pad_units": pad, "outdoor_pos": "WEST", "outdoor_depth_m": 12,
                })

    # Sweep 3: Height × outdoor
    for s in DESIGN_SPACE["storeys"]:
        for od in DESIGN_SPACE["outdoor_opts"]:
            for f in ["BAR", "L", "U"]:
                configs.append({
                    "form": f, "target_floor_area_m2": 770, "wing_width_m": 14.0,
                    "storeys": s, "corridor_type": "double", "yt_rooms": 100,
                    "pad_units": 30, "outdoor_pos": od, "outdoor_depth_m": 12,
                })

    # Sweep 4: Single-loaded
    for f in ["BAR_NS", "L"]:
        for ww in [8.0, 9.25]:
            configs.append({
                "form": f, "target_floor_area_m2": 770, "wing_width_m": ww,
                "storeys": 7, "corridor_type": "single", "yt_rooms": 100,
                "pad_units": 30, "outdoor_pos": "WEST", "outdoor_depth_m": 12,
            })

    # Generate and dedup
    options = []
    seen = set()

    for cfg in configs:
        try:
            opt = build_option(cfg)
        except Exception:
            continue
        if opt is None or (not opt["is_valid"] and not include_invalid):
            continue

        m = opt["metrics"]
        sig = (m["form"], m["total_keys"], m["yt_rooms"], m["pad_units"],
               m["storeys"], m["gia_per_key"], m["building_height_m"],
               m["west_facade_m"], m["corridor_type"], m["outdoor_pos"])
        if sig in seen:
            continue
        seen.add(sig)
        options.append(opt)

    # Sort, dedup output, assign IDs
    options.sort(key=lambda x: x["score"], reverse=True)

    for i, opt in enumerate(options[:max_options]):
        opt["rank"] = i + 1
        letter = chr(65 + i // 9)
        digit = (i % 9) + 1
        opt["id"] = f"OPT-{letter}{digit}"

    return options[:max_options]


def group_options(options: List[Dict[str, Any]]) -> Dict[str, List[str]]:
    """Group options by outcome category."""
    if not options:
        return {}
    return {
        "best_overall":   [o["id"] for o in sorted(options, key=lambda x: x["score"], reverse=True)[:5]],
        "most_rooms":     [o["id"] for o in sorted(options, key=lambda x: x["metrics"]["total_keys"], reverse=True)[:5]],
        "lowest_height":  [o["id"] for o in sorted(options, key=lambda x: x["metrics"]["building_height_m"])[:5]],
        "best_views":     [o["id"] for o in sorted(options, key=lambda x: x["metrics"]["west_facade_m"], reverse=True)[:5]],
        "lowest_cost":    [o["id"] for o in sorted(options, key=lambda x: x["metrics"]["cost_per_key_usd"])[:5]],
        "most_outdoor":   [o["id"] for o in sorted(options, key=lambda x: x["metrics"]["outdoor_total_m2"], reverse=True)[:5]],
        "most_efficient": [o["id"] for o in sorted(options, key=lambda x: abs(x["metrics"]["gia_per_key"] - 36))[:5]],
        "pad_heavy":      [o["id"] for o in sorted(options, key=lambda x: x["metrics"]["pad_units"], reverse=True)[:5]],
    }

File: studio/src/engine/generator.ts
import type {
  GenerationParams, DesignOption, FormType, CorridorType,
  OutdoorPosition, OptionMetrics, OptionGroups, ProjectId, RoomType,
} from './types'
import { generateForm } from './forms'
import { roomsPerFloor, buildFloorProgramme } from './rooms'
import { validate } from './validator'
import { scoreOption } from './scorer'
import { estimateCost } from './cost'
import { projectRevenue } from './revenue'
import { calculateAmenities } from './amenities'
import { SITE as CARLISLE_BAY_SITE } from '@/config/site'
import { YOTEL_ROOMS, YOTELPAD_UNITS } from '@/config/programme'
import { SITE as ABBEVILLE_SITE } from '@/config/abbeville/site'
import { ABBEVILLE_UNITS } from '@/config/abbeville/programme'
import { SITE as MT_BREVITOR_SITE } from '@/config/mt-brevitor/site'
import { MBE_UNITS } from '@/config/mt-brevitor/programme'
import { DEFAULT_WEIGHTS } from '@/config/scoring-weights'
import { CURATED_OPTIONS } from '@/config/curated-options'

interface SiteConfig {
  buildableArea: number
  maxCoverage: number
  maxFootprint: number
  maxHeight: number
  [key: string]: unknown
}

interface ProjectConfig {
  site: SiteConfig
  ytRooms: Record<string, RoomType>
  padUnits: Record<string, RoomType>
}

function getProjectConfig(projectId?: ProjectId): ProjectConfig {
  if (projectId === 'abbeville') {
    return {
      site: ABBEVILLE_SITE,
      ytRooms: ABBEVILLE_UNITS,
      padUnits: ABBEVILLE_UNITS,
    }
  }
  if (projectId === 'mt-brevitor') {
    return {
      site: MT_BREVITOR_SITE,
      ytRooms: MBE_UNITS,
      padUnits: MBE_UNITS,
    }
  }
  // Default: carlisle-bay
  return {
    site: CARLISLE_BAY_SITE,
    ytRooms: YOTEL_ROOMS,
    padUnits: YOTELPAD_UNITS,
  }
}

const GROUND_H = 4.5
const GROUND_H_RESIDENTIAL = 3.0  // lower ground floor for residential (no hotel lobby)
const FLOOR_H = 3.2

/** Hotel design space — Carlisle Bay and Abbeville */
const DESIGN_SPACE = {
  forms: ['BAR', 'BAR_NS', 'L', 'U', 'C'] as FormType[],
  floorAreas: [650, 770, 900, 1050],
  wingWidths: [13.6, 14.0, 16.1],
  storeys: [5, 6],
  corridors: ['double_loaded'] as CorridorType[],
  ytRooms: [95, 100, 105, 110],
  padUnits: [30, 35, 40],
  outdoor: ['WEST'] as OutdoorPosition[],
}

/**
 * Mt Brevitor Estates design space — 2-3 storey residential clusters.
 *
 * Planning envelope: max 3 storeys / 12m height (inland Barbados).
 * Single-loaded corridor = open balcony access (standard residential terrace).
 * Floor areas and wing widths sized for residential cluster blocks, not hotel.
 *
 * Heights produced:
 *   2 storeys: 3.0 + (2-1)×3.2 = 6.2 m  ✓ (<12 m)
 *   3 storeys: 3.0 + (3-1)×3.2 = 9.4 m  ✓ (<12 m)
 */
const DESIGN_SPACE_MBE = {
  forms: ['BAR', 'L', 'U'] as FormType[],
  floorAreas: [300, 450, 600, 750],        // cluster floor plates (m²)
  wingWidths: [8.0, 10.0, 12.0],           // single-loaded residential widths (m)
  storeys: [2, 3],
  corridors: ['single_loaded'] as CorridorType[],
  ytRooms: [10, 15, 20, 25, 30],           // units per cluster
  padUnits: [0],                            // no PAD split — pure residential sale
  outdoor: ['WEST'] as OutdoorPosition[],
}

function getDesignSpace(projectId?: ProjectId) {
  if (projectId === 'mt-brevitor') return DESIGN_SPACE_MBE
  return DESIGN_SPACE
}

function getGroundHeight(projectId?: ProjectId): number {
  if (projectId === 'mt-brevitor') return GROUND_H_RESIDENTIAL
  return GROUND_H
}

let optionCounter = 0

export function buildOption(params: GenerationParams, projectId?: ProjectId): DesignOption {
  const id = `opt-${++optionCounter}`
  const { site, ytRooms, padUnits } = getProjectConfig(projectId)

  // Generate form geometry
  const formResult = generateForm(params.form, params.targetFloorArea, params.wingWidth)

  // Assign storeys to wings
  const wings = formResult.wings.map(w => ({ ...w, floors: params.storeys }))

  // Calculate rooms per floor dynamically (per generator.py)
  const ytPerFloor = roomsPerFloor(wings, params.corridorType, ytRooms)
  const padPerFloor = roomsPerFloor(wings, params.corridorType, padUnits)

  // Dynamic floor allocation (per generator.py lines 60-61):
  const upperFloors = params.storeys - 1  // exclude ground
  const ytFloorCount = params.ytRooms > 0
    ? Math.min(Math.ceil(params.ytRooms / Math.max(1, ytPerFloor)), upperFloors)
    : 0
  const padFloorCount = params.padUnits > 0
    ? Math.min(Math.ceil(params.padUnits / Math.max(1, padPerFloor)), upperFloors - ytFloorCount)
    : 0

  const ytFloors = Array.from({ length: ytFloorCount }, (_, i) => i + 1)
  const actualPadFloors = Array.from({ length: padFloorCount }, (_, i) => i + 1 + ytFloorCount)

  // Actual room counts (clamped to what fits)
  const actualYtRooms = Math.min(params.ytRooms, ytFloorCount * ytPerFloor)
  const actualPadUnits = Math.min(params.padUnits, padFloorCount * padPerFloor)

  const floors = buildFloorProgramme({
    storeys: params.storeys,
    ytPerFloor: Math.round(ytPerFloor),
    padPerFloor: Math.round(padPerFloor),
    ytFloors,
    padFloors: actualPadFloors,
    footprint: formResult.footprint,
    ytRoomTypes: ytRooms,
    padRoomTypes: padUnits,
    residential: projectId === 'mt-brevitor',
  })

  // Calculate metrics
  const totalGia = floors.reduce((sum, f) => sum + f.gia, 0)
  const totalKeys = actualYtRooms + actualPadUnits
  // 130-key minimum is a YOTEL brand requirement — not applicable to residential estates
  if (projectId !== 'mt-brevitor' && totalKeys < 130) {
    throw new Error(`Below 130-key minimum (got ${totalKeys})`)
  }
  // Use project-specific ground floor height (residential = 3.0m, hotel = 4.5m)
  const height = getGroundHeight(projectId) + (params.storeys - 1) * FLOOR_H
  const coverage = formResult.footprint / site.buildableArea
  const outdoorTotal = params.outdoorPosition === 'BOTH' ? 660 + 80 :
                       params.outdoorPosition === 'ROOFTOP' ? 80 : 660

  const metrics: OptionMetrics = {
    totalKeys,
    yotelKeys: actualYtRooms,
    padUnits: actualPadUnits,
    gia: totalGia,
    giaPerKey: totalGia / Math.max(1, totalKeys),
    footprint: formResult.footprint,
    coverage,
    buildingHeight: height,
    westFacade: formResult.westFacade,
    outdoorTotal,
    costPerKey: 0,
    tdc: 0,
    corridorType: params.corridorType,
    form: params.form,
    amenityScore: 0,
  }

  // Cost
  const cost = estimateCost(metrics)
  metrics.tdc = cost.total
  metrics.costPerKey = cost.perKey

  // Revenue
  const revenue = projectRevenue(actualYtRooms, actualPadUnits, 5)

  // Amenities
  const amenities = calculateAmenities(
    metrics.totalKeys,
    metrics.footprint,
    metrics.buildingHeight,
    80, // rooftop base from programme
    metrics.westFacade,
    params.form,
  )
  metrics.amenityScore = amenities.amenityScore

  // Validate against project-specific site constraints
  const validation = validate(metrics, wings, undefined, projectId)

  // Score
  const [score, scoringBreakdown] = scoreOption(metrics, DEFAULT_WEIGHTS)

  return {
    id, form: params.form, params, wings, floors, metrics,
    amenities, cost, revenue, score, scoringBreakdown, validation,
  }
}

/**
 * Find optimal YT/PAD split for a given form configuration.
 * Tests splits from 60% to 85% YOTEL (in 5% steps) and picks
 * the one with highest yield on cost.
 *
 * Not applicable to Mt Brevitor — it is a pure residential sale estate
 * with no YOTEL/PAD split to optimise (returns null immediately).
 */
function optimizeUnitMix(
  form: FormType, area: number, width: number, storeys: number, projectId?: ProjectId
): { ytRooms: number; padUnits: number; yieldOnCost: number } | null {
  // Residential estates have no hotel YT/PAD mix to optimise
  if (projectId === 'mt-brevitor') return null

  let bestYield = -1
  let bestSplit = { ytRooms: 100, padUnits: 30, yieldOnCost: 0 }
  const { ytRooms: ytRoomConfig } = getProjectConfig(projectId)

  // Total capacity estimate
  const formResult = generateForm(form, area, width)
  const wings = formResult.wings.map(w => ({ ...w, floors: storeys }))
  const roomsFloor = roomsPerFloor(wings, 'double_loaded', ytRoomConfig)
  const upperFloors = storeys - 1
  const maxRooms = Math.floor(roomsFloor * upperFloors)

  for (let ytPct = 60; ytPct <= 85; ytPct += 5) {
    const yt = Math.round(maxRooms * ytPct / 100)
    const pad = maxRooms - yt

    if (yt < 50 || pad < 10 || yt + pad < 80) continue

    try {
      const params: GenerationParams = {
        form, targetFloorArea: area, wingWidth: width,
        storeys, corridorType: 'double_loaded' as const,
        ytRooms: yt, padUnits: pad, outdoorPosition: 'WEST' as const,
      }
      const opt = buildOption(params, projectId)
      if (!opt.validation.isValid) continue

      const yoc = opt.cost.total > 0 ? opt.revenue.stabilisedNoi / opt.cost.total : 0
      if (yoc > bestYield) {
        bestYield = yoc
        bestSplit = { ytRooms: yt, padUnits: pad, yieldOnCost: yoc }
      }
    } catch {
      continue
    }
  }

  return bestYield > 0 ? bestSplit : null
}

export function generateAll(maxOptions: number = 50, projectId?: ProjectId): DesignOption[] {
  optionCounter = 0

  // Select the design space for this project
  const ds = getDesignSpace(projectId)

  // ── Phase 1: Generate curated Architect's Options (Carlisle Bay only) ──────
  // CURATED_OPTIONS are calibrated specifically for the Carlisle Bay hotel site.
  // Skip entirely for Abbeville and Mt Brevitor — they get parametric-only results.
  const curated: DesignOption[] = []
  if (!projectId || projectId === 'carlisle-bay') {
    for (const cfg of CURATED_OPTIONS) {
      try {
        const opt = buildOption(cfg.params, projectId)
        opt.curatedId = cfg.id
        opt.curatedName = cfg.name
        opt.curatedConcept = cfg.concept
        // Always include curated designs — they're architect-approved.
        // Warnings are acceptable; only skip if buildOption() throws (fatal constraint violation).
        curated.push(opt)
      } catch {
        // Skip curated options that fail generation (e.g. below 130-key minimum)
        console.warn(`Curated design "${cfg.name}" (${cfg.id}) failed generation`)
      }
    }
  }

  // Sort curated: recommended first, then by score descending
  curated.sort((a, b) => {
    const aCfg = CURATED_OPTIONS.find(c => c.id === a.curatedId)
    const bCfg = CURATED_OPTIONS.find(c => c.id === b.curatedId)
    const aRec = aCfg?.recommended ? 1 : 0
    const bRec = bCfg?.recommended ? 1 : 0
    if (aRec !== bRec) return bRec - aRec
    return b.score - a.score
  })

  // ── Phase 2: Parametric sweep ───────────────────────────────────────
  const sweepOptions: DesignOption[] = []

  for (const form of ds.forms) {
    for (const area of ds.floorAreas) {
      for (const width of ds.wingWidths) {
        for (const storeys of ds.storeys) {
          for (const corridor of ds.corridors) {
            for (const yt of ds.ytRooms) {
              for (const pad of ds.padUnits) {
                const params: GenerationParams = {
                  form, targetFloorArea: area, wingWidth: width,
                  storeys, corridorType: corridor,
                  ytRooms: yt, padUnits: pad, outdoorPosition: 'WEST',
                }
                try {
                  const opt = buildOption(params, projectId)
                  if (opt.validation.isValid) sweepOptions.push(opt)
                } catch {
                  // Skip invalid parameter combinations
                }
              }
            }
          }
        }
      }
    }
  }

  // Add optimized unit mix options for hotel projects (not residential estates)
  for (const form of ds.forms) {
    for (const area of ds.floorAreas) {
      for (const width of ds.wingWidths) {
        for (const storeys of ds.storeys) {
          const optimal = optimizeUnitMix(form, area, width, storeys, projectId)
          if (optimal) {
            try {
              const params: GenerationParams = {
                form, targetFloorArea: area, wingWidth: width,
                storeys, corridorType: ds.corridors[0],
                ytRooms: optimal.ytRooms, padUnits: optimal.padUnits,
                outdoorPosition: 'WEST' as const,
              }
              const opt = buildOption(params, projectId)
              if (opt.validation.isValid) sweepOptions.push(opt)
            } catch {
              // Skip invalid optimized combinations
            }
          }
        }
      }
    }
  }

  // Deduplicate sweep by output signature (form + keys + yt/pad split + storeys)
  const curatedSigs = new Set(curated.map(o =>
    `${o.form}-${o.metrics.totalKeys}-${o.metrics.yotelKeys}-${o.metrics.padUnits}-${Math.round(o.metrics.buildingHeight)}`
  ))
  const seen = new Set<string>(curatedSigs)
  const unique = sweepOptions.filter(o => {
    const sig = `${o.form}-${o.metrics.totalKeys}-${o.metrics.yotelKeys}-${o.metrics.padUnits}-${Math.round(o.metrics.buildingHeight)}`
    if (seen.has(sig)) return false
    seen.add(sig)
    return true
  })

  // Sort sweep by score descending
  unique.sort((a, b) => b.score - a.score)

  // Round-robin across form types to ensure diversity in sweep results
  const byForm = new Map<string, DesignOption[]>()
  for (const o of unique) {
    const arr = byForm.get(o.form) ?? []
    arr.push(o)
    byForm.set(o.form, arr)
  }

  const sweepResult: DesignOption[] = []
  const formIndices = new Map<string, number>()
  for (const form of ds.forms) formIndices.set(form, 0)

  const sweepLimit = maxOptions - curated.length
  let added = true
  while (sweepResult.length < sweepLimit && added) {
    added = false
    for (const form of ds.forms) {
      if (sweepResult.length >= sweepLimit) break
      const idx = formIndices.get(form)!
      const formOpts = byForm.get(form) ?? []
      if (idx < formOpts.length) {
        sweepResult.push(formOpts[idx])
        formIndices.set(form, idx + 1)
        added = true
      }
    }
  }

  // Sort sweep by score
  sweepResult.sort((a, b) => b.score - a.score)

  // Return curated first, then sweep
  return [...curated, ...sweepResult]
}

export function groupOptions(options: DesignOption[]): OptionGroups {
  const sorted = [...options]
  return {
    best_overall: sorted.slice(0, 5).map(o => o.id),
    most_rooms: [...sorted].sort((a, b) => b.metrics.totalKeys - a.metrics.totalKeys).slice(0, 3).map(o => o.id),
    lowest_height: [...sorted].sort((a, b) => a.metrics.buildingHeight - b.metrics.buildingHeight).slice(0, 3).map(o => o.id),
    best_views: [...sorted].sort((a, b) => b.metrics.westFacade - a.metrics.westFacade).slice(0, 3).map(o => o.id),
    lowest_cost: [...sorted].sort((a, b) => a.metrics.costPerKey - b.metrics.costPerKey).slice(0, 3).map(o => o.id),
    most_outdoor: [...sorted].sort((a, b) => b.metrics.outdoorTotal - a.metrics.outdoorTotal).slice(0, 3).map(o => o.id),
    most_efficient: [...sorted].sort((a, b) => a.metrics.giaPerKey - b.metrics.giaPerKey).slice(0, 3).map(o => o.id),
    pad_heavy: [...sorted].sort((a, b) => b.metrics.padUnits - a.metrics.padUnits).slice(0, 3).map(o => o.id),
  }
}
